libclamav/ole2_extract.c

```diff
@@ -1756,6 +2024,423 @@ ole2_read_header(int fd, ole2_header_t *hdr)
+
+    buffer = calloc(bufLen, 1);
+    if (NULL == buffer) {
+        perror("calloc");
+        goto done;
+    }
+
+    memcpy(buffer, verifier->salt, verifier->salt_size);
+
+    /*Convert to UTF16-LE*/
+    for (i = 0; i < (uint32_t)strlen(password); i++) {
+        buffer[verifier->salt_size + (i * 2)] = password[i];
+    }
+
+    cl_sha1(buffer, bufLen, sha1Dst, NULL);
+
+    for (i = 0; i < ITER_COUNT; i++) {
+        uint32_t eye = ole2_endian_convert_32(i);
+
+        memcpy(sha1, &eye, sizeof(eye));
+        cl_sha1(sha1, SHA1_HASH_SIZE + sizeof(uint32_t), sha1Dst, NULL);
+    }
+
+    memset(&(sha1Dst[SHA1_HASH_SIZE]), 0, sizeof(uint32_t));
+
+    cl_sha1(sha1Dst, SHA1_HASH_SIZE + sizeof(uint32_t), sha1Dst, NULL);
+
+    memset(buf1, 0x36, sizeof(buf1));
+    for (i = 0; i < SHA1_HASH_SIZE; i++) {
+        buf1[i] = buf1[i] ^ sha1Dst[i];
+    }
+
+    //now sha1 buf1
+    cl_sha1(buf1, sizeof(buf1), doubleSha, NULL);
+
+    memset(buf2, 0x5c, sizeof(buf2));
+    for (i = 0; i < SHA1_HASH_SIZE; i++) {
+        buf2[i] = buf2[i] ^ sha1Dst[i];
+    }
+
+    cl_sha1(buf2, sizeof(buf2), &(doubleSha[SHA1_HASH_SIZE]), NULL);
+
+    ret = 0;
+done:
+    FREE(buffer);
+
+    if (0 == ret) {
+        memcpy(key, doubleSha, keyLength);
+    }
+
+    return ret;
+}
+
+static void aes_128ecb_decrypt(const unsigned char *in, size_t length, unsigned char *out, const uint8_t *key, unsigned key_n)
+{
+
+    unsigned long rk[RKLENGTH(128)];
+    int nrounds;
+    size_t i;
+
+    nrounds = rijndaelSetupDecrypt(rk, (const unsigned char *)key, key_n * 8);
+    if (!nrounds) {
+        cli_errmsg("ole2: nrounds NULL\n");
+    } else {
+        for (i = 0; i < length; i += 16) {
+            rijndaelDecrypt(rk, nrounds, &(in[i]), &(out[i]));
+        }
+    }
+}
+
+/*
+ * Returns true if it is actually encrypted with the key.
+ *
+ * Details here
+ * https://docs.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/e5ad39b8-9bc1-4a19-bad3-44e6246d21e6
+ */
+static bool verify_key(uint8_t key[16], encryption_verifier_t *verifier)
+{
+#define AES_VERIFIER_HASH_LEN 32
+
+    bool bRet = false;
+    uint8_t sha[SHA1_HASH_SIZE];
+    const uint32_t encrypted_verifier_hash_len = AES_VERIFIER_HASH_LEN;
+    uint8_t decrypted[encrypted_verifier_hash_len];
+
+    aes_128ecb_decrypt(verifier->encrypted_verifier, sizeof(verifier->encrypted_verifier),
+                       decrypted, key, 16);
+
+    cl_sha1(decrypted, sizeof(verifier->encrypted_verifier), sha, NULL);
+
+    aes_128ecb_decrypt(verifier->encrypted_verifier_hash, verifier->verifier_hash_size,
+                       decrypted, key, 16);
+
+    bRet = (0 == memcmp(sha, decrypted, verifier->verifier_hash_size));
+
+    return bRet;
+}
+
+//https://docs.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/dca653b5-b93b-48df-8e1e-0fb9e1c83b0f
+//https://learn.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/2895eba1-acb1-4624-9bde-2cdad3fea015
```

Comment:
This style commend block would be nice, along with an explanation as to what each link is for, in case Microsoft moves the links and the links die, and also so that I don't have to click them to know if I really want to read them. 

```c
/**
 * @brief 
 * 
 * @param ctx 
 * @param headerPtr 
 * @param encryptionKey 
 * @return true 
 * @return false 
 */
```